libc-top-half/musl/src/string/strspn_simd.c

```diff
@@ -0,0 +1,175 @@
+#if defined(__wasm_simd128__) && defined(__wasilibc_simd_string)
+// Skip Clang 19 and Clang 20 which have a bug (llvm/llvm-project#146574)
+// which results in an ICE when inline assembly is used with a vector result.
+#if __clang_major__ != 19 && __clang_major__ != 20
+
+#include <stdint.h>
+#include <string.h>
+#include <wasm_simd128.h>
+
+#if !defined(__wasm_relaxed_simd__) || !defined(__RELAXED_FN_ATTRS)
+#define wasm_i8x16_relaxed_swizzle wasm_i8x16_swizzle
+#endif
+
+// SIMDized check which bytes are in a set (Geoff Langdale)
+// http://0x80.pl/notesen/2018-10-18-simd-byte-lookup.html
+
+typedef struct {
+  __u8x16 lo;
+  __u8x16 hi;
+} __wasm_v128_bitmap256_t;
+
+__attribute__((always_inline))
+static void __wasm_v128_setbit(__wasm_v128_bitmap256_t *bitmap, int i) {
+  uint8_t hi_nibble = (uint8_t)i >> 4;
+  uint8_t lo_nibble = (uint8_t)i & 0xf;
+  bitmap->lo[lo_nibble] |= (uint8_t)((uint32_t)1 << (hi_nibble - 0));
+  bitmap->hi[lo_nibble] |= (uint8_t)((uint32_t)1 << (hi_nibble - 8));
+}
+
+__attribute__((always_inline))
+static v128_t __wasm_v128_chkbits(__wasm_v128_bitmap256_t bitmap, v128_t v) {
+  v128_t hi_nibbles = wasm_u8x16_shr(v, 4);
+  v128_t bitmask_lookup = wasm_u8x16_const(1, 2, 4, 8, 16, 32, 64, 128,  //
+                                           1, 2, 4, 8, 16, 32, 64, 128);
```

Comment:
These are both `-128` in the [original algorithm], no?

[original algorithm]: https://web.archive.org/web/20250409223546/http://0x80.pl/notesen/2018-10-18-simd-byte-lookup.html

Reply:
Yes. In two's complement, `int8(-128)` is equal to `uint8(128)`. In interpreting the algorithm, I found unsigned clearer, so used that.

I guess Intel lacks the unsigned version of `_mm_setr_epi8` intrinsic (which would make to the same instruction anyways…)?

What do you think is best, a comment explaining this, or `wasm_i8x16_const`?